bitbake-setup: allow passing SRC_URI as arg.config

js731ca opened the pull request:
On CI setups with a transient build-directory it would be desirable to skip the global-settings and settings configuration when dealing with a "downstream" config registry, and have a one-line build-setup command.

After some minor refactoring, this would now be possible:
```
/bitbake/bin/bitbake-setup --global-settings /dev/null init --top-dir-prefix /tmp/bitbake-setup --top-dir-name acme 'git://github.com/acme/meta-acme;branch=scarthgap-bitbake-setup;protocol=https;rev=scarthgap-bitbake-setup' acme_5.0.x gizmo --non-interactive
```
E.g.: disabling the global config, passing the registry SRC_URI in directly, and from it select a conf.json with its bitbake-setup.

rpurdie committed:
bitbake-setup: Allow local registry paths

It is useful for bitbake-setup to support local paths without access through
the fetcher so that internal data to the bitbake repository can be used as
a default.

Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

rpurdie committed:
bitbake-setup: Switch to internal default registry files

Switch the url to be the default internal registry rather than a private
repo which was intended for testing.

Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

rpurdie committed:
Add default registry config for OpenEmbedded's nodistro and Yocto Project's poky

Rather than forcing every user to have to access a separate repository
for data for some key defaults, add those for nodistro and poky. This
gives us some commonly used setups and also something to test against too.

We will need to come up with a criteria for adding these, most likely based
on community usage/demand with the OE TSC governing that policy.

Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

kanavin committed:
bitbake-setup: add the initial implementation

Preamble
========

The latest iteration of this patchset is available at
https://github.com/kanavin/bitbake
I recommend taking the patches from there to ensure that
you are not trying out outdated code.

For the rationale and design guidelines please see this message:
https://lists.openembedded.org/g/openembedded-architecture/message/1913

Left out for now but will be done later:

- official configuration repository

- documentation

Amble *scratch* HOWTO
=====================

1. If you don't know where to start, run 'bitbake-setup init'.

Bitbake-setup will ask a few questions about available configuration choices and set up a build.

Note: 'init' sub-command can also take a path or a URL with a configuration file directly.
You can see how those files look like here:
https://github.com/kanavin/bitbake-setup-configurations

2. You can then source the bitbake environment and run bitbake to perform builds as usual:

$ . /home/alex/bitbake-builds/yocto-master-options-poky-distro_poky-machine_qemux86-64/build/init-build-env

Also, subsequent status/update commands will not require a separate --build-dir argument telling
bitbake-setup where the build is.

3. To check if the build configuration needs to be updated, run:

===
$ bin/bitbake-setup status
...
Configuration in /home/alex/bitbake-builds/poky-alex/ has not changed.
===

If the configuration has changed, you will see the difference as a diff.
...
  -                "rev": "akanavin/sstate-for-all"
  +                "rev": "akanavin/bitbake-setup-testing"
...

If the configuration has not changed, but layer revisions referred to it have (for example
if the configuration specifies a tip of a branch), you will see that too:

===
...
Layer repository git://git.yoctoproject.org/poky-contrib checked out into /home/alex/builds/poky-alex/layers/poky updated revision akanavin/sstate-for-all from 6b842ba55f996b27c900e3de78ceac8cb3b1c492 to aeb73e29379fe6007a8adc8d94c1ac18a93e68de
===

4. If the configuration has changed, you can bring it in sync with:

$ bin/bitbake-setup update

Note that it will also rename/preserve the existing build/conf directory, and print changes
in bitbake configuration (diff of content of build/conf/) if that has changed. I can't
at the moment think of anything more clever that is also not much more brittle or complex
to implement, but open to suggestions.

Terminology
===========

- 'top directory' means the place under which bitbake-setup reads and
writes everything. bitbake-setup makes a promise to not touch anything outside of
that, unless otherwise directed to by entries in settings (currently
there is one such setting for fetcher downloads for layers and config
registries). Top directory can be selected by an environment variable, a command line option,
or otherwise assumed to be ~/bitbake-builds/. If BBPATH is in environment
(e.g. we are in a bitbake environment), then the top directory is
deduced from that and doesn't need to be specified by hand.

- 'settings' means bitbake-setup operational parameters that are
global to all builds under a top directory. E.g. the location of
configuration registry, or where the bitbake fetcher should place the
downloads (DL_DIR setting). Settings are stored in a .conf file in ini
format just under the top directory.

- 'build' means a tree structure set up by 'bitbake-setup init',
consisting of, at least, a layers checkout, and a bitbake
build. It maps 1:1 to the json data it was constructed from, which is
called 'build configuration'. Build configurations are constructed from
generic configurations that may involve making one or more choices
about available options in them. Generic configurations are files, URLs
or are obtained from git repositories called 'config
registries', in which case they can be listed with 'bitbake-setup
list'. There can be multiple 'builds' under a top directory. Here are
two example generic configurations that showcase this:
https://github.com/kanavin/bitbake-setup-configurations/blob/main/yocto-master-options.conf.json
https://github.com/kanavin/bitbake-setup-configurations/blob/main/yocto-master-nested-configs.conf.json

- 'bitbake-setup status' will tell if a build is in sync with
the generic configuration it was made from. 'bitbake-setup update' will bring a build
in sync with a configuration if needed.

- 'bitbake build' means a particular sub-tree inside a build that
bitbake itself operates on, e.g. what is set in BBPATH/BUILDDIR
by oe-init-build-env. conf/* in that tree is 'bitbake configuration'.
Bitbake configurations are constructed from templates and fragments,
with existing mechanisms provided by oe-core. The configuration file
format is specified such that other mechanisms to set up a
bitbake build can be added; there was a mention of ability to specify
local.conf content and a set of layers directly in a configuration. I
think that scales poorly compared to templates and fragments, but I
made sure alternative ways to configure a bitbake build are possible
to add in the future :)

- 'source override' is a json file that can be used to modify revisions
and origins of layers that need to be checkout into a build (e.g.
when master branches need to be changed to master-next for purposes
of testing). Such a file is specified with a command-line option to 'init'
and an example can be seen here:
https://github.com/kanavin/bitbake-setup-configurations/blob/main/yocto-master-next.override.json

This commit includes fixes by
Ryan Eatmon <reatmon@ti.com> https://github.com/kanavin/bitbake/pull/1
Gyorgy Sarvari <skandigraun@gmail.com> https://github.com/kanavin/bitbake/pull/2
Johannes Schneider <johannes.schneider@leica-geosystems.com> https://github.com/kanavin/bitbake/pull/3 https://github.com/kanavin/bitbake/pull/5

Signed-off-by: Alexander Kanavin <alex@linutronix.de>
Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

kanavin committed:
bitbake-setup: add tests to bitbake-selftest

Run like this:

alex@Zen2:/srv/work/alex/bitbake$ bin/bitbake-selftest -v bb.tests.setup
test_setup (bb.tests.setup.BitbakeSetupTest.test_setup) ... ok

----------------------------------------------------------------------
Ran 1 test in 9.223s

OK

The test does a basic run-through of init, then status/update
on an unchanged configuration, then status/update on a
configuration changed via new commits to the test layer,
then status/update on configuration changed via the top
level json config file.

Note that nothing whatsoever is fetched from the network;
the test relies entirely on synthetic data contained inside
itself, including minimal stubs for oe-setup-build and
bitbake-config-build. This data is used to create temporary
git repositories then clone them via local filesystem URIs.

Later on this can be supplemented by an oe-selftest that
tests bitbake-setup against real config files in the
official configuration repository and real layers,
templates and fragments.

Signed-off-by: Alexander Kanavin <alex@linutronix.de>
Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

kanavin committed:
bitbake-setup: add 'install-buildtools' command

This basically calls install-buildtools from oe-core/poky, but
it ensures via command line parameters that the installation
location is stable and the downloads are preserved for reproducibility:

$ bin/bitbake-setup install-buildtools
Loading settings from /home/alex/bitbake-builds/bitbake-setup.conf

======
Buildtools archive is downloaded into /home/alex/bitbake-builds/yocto-master-testing/buildtools-downloads/20250319141333 and its content installed into /home/alex/bitbake-builds/yocto-master-testing/buildtools

... (output from install-buildtools script)
======

It also detects when buildtools are already installed, and will direct
users what to do:
======
alex@Zen2:/srv/work/alex/bitbake$ bin/bitbake-setup install-buildtools
Loading settings from /home/alex/bitbake-builds/bitbake-setup.conf

Buildtools are already installed in /home/alex/bitbake-builds/yocto-master-testing/buildtools.
If you wish to use them, you need to source the the environment setup script e.g.
$ . /home/alex/bitbake-builds/yocto-master-testing/buildtools/environment-setup-x86_64-pokysdk-linux
You can also re-run bitbake-setup install-buildtools with --force option to force a reinstallation
======

This commits includes fixes by Gyorgy Sarvari <skandigraun@gmail.com>
https://github.com/kanavin/bitbake/pull/2

Signed-off-by: Alexander Kanavin <alex@linutronix.de>
Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

kanavin committed:
cooker: adjust the error message about missing websockets module to mention bitbake-setup

Signed-off-by: Alexander Kanavin <alex@linutronix.de>
Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

ycongal-smile committed:
bitbake-setup: dash support for init-build-env script

Being minimalist, dash does not support the (non-POSIX) feature of
passing an argument while sourcing a script. Like in
  . <some path>/oe-init-build-env <build dir>

With dash, one must use:
  set <build dir>       # puts <build dir> in $1
  cd <some path>
  . ./oe-init-build-env # can only be called from its directory in dash

To do this:
* Instead of a symlink to oe-init-build-env, keep a symlink to the
  directory containing it (called "oe-init-build-env-dir")
* Generate a init-build-env script that dash can source using the above
  snippet.

Signed-off-by: Yoann Congal <yoann.congal@smile.fr>

rpurdie committed:
bitbake-setup: Improve the already initialized test

If the directory already exists but hasn't been setup, the current test
can fail so improve it.

Signed-off-by: Richard Purdie <richard.purdie@linuxfoundation.org>

ycongal-smile committed:
bitbake-setup: suggest "." instead of "source"

"." is in POSIX standard[0], whereas "source" is only supported in more
feature-full shells (bash, zsh, ...)

[0]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#dot

Signed-off-by: Yoann Congal <yoann.congal@smile.fr>

kanavin committed:
bitbake-setup: add support for specifying branches in repo checkouts

Previously bitbake-setup was checking out 'detached commits' using
fetcher's nobranch feature, as that is the only option when only a revision is in the config.

Branches are optional, but beneficial, as

- checkout directory will be on a branch, making it easier for users
to understand where they are if they need to make changes (also
bitbake will print branch information instead of saying 'HEAD:sha').

- supply chain security! Enforcing a branch means any specified revision
has to be on it, and no one can sneak in (accidentally or deliberately!)
some dangling commit, or something from their private branch in the same repo.

Signed-off-by: Alexander Kanavin <alex@linutronix.de>

kanavin committed:
bitbake-setup: improve robustness of loading/writing settings

Particularly:

- ensure global settings command line argument is always expanded to
full path
- ensure any errors that happen when loading settings are reported
at that point, otherwise we get an empty dictionary and cryptic
key errors later

Signed-off-by: Alexander Kanavin <alex@linutronix.de>

kanavin commented:
Same here, please rebase and drop my obsolete commits from the PR.

ycongal-smile committed:
bitbake-setup: allow using {THISDIR}/my-layer

This implement the ability to use "{THISDIR}/my-layer" in the
"bb-layers" list. "{THISDIR}" is remplaced by the directory containing
the configuration file.

In small projects, we try to keep the setup a simple as possible: a
single git repo containing both the build confguration (e.g.
a bitbake-setup configuration file) and the meta layer with project
recipes/machine/distro.

This change allows this kind of setup:
├── meta-my-project/ # the project layer
└── my-project.conf.json   # the bb-setup configuration file

by writing, in my-project.conf.json:
  "bitbake-setup": {
    "configurations": [{
      "bb-layers": [
        "{THISDIR}/meta-my-project"

Note: in this case meta-my-project is not present as a "source", so, not
handled by bb-setup update/status. It is expected of the user to handle
this on their own (is our case, a simple git workflow).

Signed-off-by: Yoann Congal <yoann.congal@smile.fr>

kanavin committed:
lib/bb/tests/setup.py: define test parameters in a single dictionary

This makes maintaining and extending them easier.

Signed-off-by: Alexander Kanavin <alex@linutronix.de>

kanavin committed:
lib/bb/tests/setup.py: unset BBPATH to ensure isolation from the existing bitbake environment

bitbake-setup deduces top directory from BBPATH, which, if set, interferes with
the tests' own setup.

Signed-off-by: Alexander Kanavin <alex@linutronix.de>

kanavin commented:
Other than rebasing, I think this does need to be split into two separate MRs: one for the settings refactor and one for passing in the registry URI on the command line. Additionally the settings refactor in itself needs to be split in several commits: one that renames the function, another that splits site.conf creation into a function, yet another that splits actual file write into a function, and only then rearranging the logic. Otherwise I simply can't tell if the logic in creating the needed files as needed is still accurate, especially when done implicitly as part of 'init' sequence.

kanavin commented:
Regarding passing in the registry URI, I'd really rather make it an explicit option, than an implicit positional argument - those are meant to choose from configuration options, not to override settings, and registry URI is a setting.

Which, in turn, raises the other question: is running bitbake-setup more than once to set where the registry is so bad? You can chain these invocations with shell's && and that's only slightly longer than a monstrous single invocation, and probably more readable too.

We'll probably have more settings in the future, there's already a setting for where downloads go, should they all come with their own command line options? I'm not so sure.

paulgortmaker committed:
bitbake-setup: clarify that default answer to prompts is no

It is common practice to put the default choice in upper case for
yes/no interactive prompts, so that when people just hit enter, they
know what they are getting.  An example that linux users are probably
familiar with is "sensors-detect" from the "lm-sensors" package.

Unify all the prompts to be the same and indicate that the default
answer from hitting enter is a no with an upper case N.  No functional
changes.

Signed-off-by: Paul Gortmaker <paulg@kernel.org>

kanavin committed:
bitbake: registry: make a separate configuration for poky with sstate mirror

Using sstate mirror places much higher requirements on both network robustness
and network bandwidth than just fetching source code. When the network doesn't meet
those, the user experience can be very frustrating as errors can be cryptic and
intermittent, or bitbake would just seemingly do nothing for a long time.

Let's make sstate an opt-in, and provide a caution for using it. When it does
work, it works wonderfully and does accelerate builds significantly.

Signed-off-by: Alexander Kanavin <alex@linutronix.de>

kanavin committed:
bitbake-setup: add support for skipping a fragment selection

In autobuilder testing a use case arised where
- the available choices in configuration file for choosing a machine are incomplete
- putting every possible machine choice into that configuration is undesirable/not possible
- autobuilder code can write a machine selection into the bitbake config
after the fact.

So this --skip-selection option is intended for advanced users that know what they're doing
and is generally not recommended as it requires manually tweaking the bitbake config to
make it usable.

Signed-off-by: Alexander Kanavin <alex@linutronix.de>

js731ca committed:
bitbake-setup: add '--setting' argument to 'init'

Add an optional argument to the init command that lets a user directly
pass in section:key:value triplets for the settings file that will be
initialized in the top-dir.

With this it is now possible to directly set e.g. a registry to fetch
the conf.json from when 'init'-ializing a build setup; avoiding
separate bitbake-setup invocations to create the top-dir configuration
first.

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented:
> I think this does need to be split into two separate MRs: one for the settings refactor and one for passing in the registry URI on the command line. Additionally the settings refactor in itself needs to be split in several commits: one that renames the function, another that splits site.conf creation into a function, yet another that splits actual file write into a function, and only then rearranging the logic. Otherwise I simply can't tell if the logic in creating the needed files as needed is still accurate, especially when done implicitly as part of 'init' sequence.

Done. The refactoring is now at #13 (and not directly needed by the next version of this MR)

js731ca commented:
> Regarding passing in the registry URI, I'd really rather make it an explicit option, than an implicit positional argument - those are meant to choose from configuration options, not to override settings, and registry URI is a setting.

Ok, then this 469e0e9 would have "abused" the configuration selection... i'll rework/drop it. 

> Which, in turn, raises the other question: is running bitbake-setup more than once to set where the registry is so bad? You can chain these invocations with shell's && and that's only slightly longer than a monstrous single invocation, and probably more readable too.
> 

Here your sales pitch comes back to haunt you:
"Producing a Complete Linux System With a Single Command (and Configuration File) With Yocto - Alexander Kanavin, Linutronix" ;-)

Currently we have three configuration files (global, settings and the actual.conf.json) and (in this case at least) three commands.

> We'll probably have more settings in the future, there's already a setting for where downloads go, should they all come with their own command line options? I'm not so sure.

This gave me the idea for a more elegant(?) solution, how about: d0d7ea13 - which now allows a one-liner like:
```
/bitbake/bin/bitbake-setup \
    --global-settings /dev/null \
    init \
    --top-dir-prefix /tmp/bitbake-setup \
    --top-dir-name acme 
    --setting default registry 'git://github.com/acme/meta-acme;branch=scarthgap-bitbake-setup;protocol=https;rev=scarthgap-bitbake-setup' \
   acme_5.0.x \
   gizmo \
   --non-interactive
```

kanavin commented on bin/bitbake-setup:
```diff
@@ -853,6 +859,11 @@ def main():
+
+        if args.setting:
+            for section, key, value in args.setting:
+                change_settings(args.top_dir, {section:{key:value}})
```
change_setting() actually writes out the settings, I'd say --setting needs to have effect for only the invocation it's used in and not do any file modifications.

For consistency, there should also be --global-setting. Then you wont have to set --global-settings-file to /dev/null (ugh!).

kanavin commented:
I'm now doing a comprehensive settings rework that will likely incorporate ideas from both this and #13. Please hold off any further work until that is ready and pushed (in all likelihood, it might obsolete both PRs).

kanavin commented:
I pushed the settings rework, I think as it (among other things) adds a --setting option this can be closed?

js731ca commented:
thanks for picking up and running with the `--settings` idea!
yes - 231738704a5d2cbd4d435226c9b4d2219f1f7ce7 fully covers the usecase

kanavin commented:
> thanks for picking up and running with the `--settings` idea! yes - [2317387](https://github.com/kanavin/bitbake/commit/231738704a5d2cbd4d435226c9b4d2219f1f7ce7) fully covers the usecase

Please also check if #13 is obsolete.
